src/gutter_tree.cpp

```diff
@@ -0,0 +1,598 @@
+
+  return dst;
+}
+
+// copy two serailized updates between two locations
+inline void GutterTree::copy_serial(char *src, char *dst) {
+  memcpy(dst, src, serial_update_size);
+}
+
+/*
+ * Load a key from a given location
+ */
+inline node_id_t GutterTree::load_key(char *location) {
+  node_id_t key;
+  memcpy(&key, location, sizeof(node_id_t));
+  return key;
+}
+
+/*
+ * Helper function which determines which child we should flush to
+ */
+inline uint32_t which_child(node_id_t key, node_id_t min_key, node_id_t max_key, uint16_t options) {
+  node_id_t total = max_key - min_key + 1;
+  double div = (double)total / options;
+
+  uint32_t larger_kids = total % options;
+  uint32_t larger_count = larger_kids * ceil(div);
+  node_id_t idx = key - min_key;
+
+  if (idx >= larger_count)
+    return ((idx - larger_count) / (int)div) + larger_kids;
+  else
+    return idx / ceil(div);
+}
+
+/*
+ * Perform an insertion to the buffer-tree
+ * Insertions always go to the root
+ */
+insert_ret_t GutterTree::insert(const update_t &upd) {
+  // printf("inserting to gutter_tree . . . \n");
+  
+  // first calculate which of the roots we're inserting to
+  Node key = upd.first;
+  buffer_id_t r_id = which_child(key, 0, num_nodes-1, fanout); // TODO: Here we are assuming that num_nodes >= fanout
+  BufferControlBlock *root = buffers[r_id];
+  // printf("Insertion to buffer %i of size %llu\n", r_id, root->size());
+
+  // perform the write and block if necessary
+  while (true) {
+    std::unique_lock<std::mutex> lk(BufferControlBlock::buffer_ready_lock);
+    BufferControlBlock::buffer_ready.wait(lk, [root, this]{return (root->size() < buffer_size + page_size);});
+    if (root->size() < buffer_size + page_size) {
+      lk.unlock();
+      root->lock_rw(); // ensure that a worker isn't flushing this node.
+      serialize_update(GutterTree::cache + root->size() + root->offset(), upd);
+      root->set_size(root->size() + serial_update_size);
+      // printf("Did an insertion. Root %i size now %lu\n", r_id, root->size());
+      break;
+    }
+    lk.unlock();
+  }
+
+  // if the buffer is full enough, push it to the flush_queue
+  if (root->size() > buffer_size && root->size() - serial_update_size <= buffer_size) {
+    BufferFlusher::queue_lock.lock();
+    BufferFlusher::flush_queue.push(r_id);
+    BufferFlusher::queue_lock.unlock();
+    BufferFlusher::flush_ready.notify_one();
+    // printf("Added to %i to flush queue\n", r_id);
+  }
+  root->unlock_rw(); // no longer need root lock
+}
+
+/*
+ * Function for perfoming a flush anywhere in the tree agnostic to position.
+ * this function should perform correctly so long as the parameters are correct.
+ *
+ * IMPORTANT: after perfoming the flush it is the caller's responsibility to reset
+ * the number of elements in the buffer and associated pointers.
+ *
+ * IMPORTANT: Unless we add more flush_buffers only a single flush at each level may occur 
```

Comment:
is this still pertinent?

Reply:
Last point is slightly wrong. I'll update that